### ISSUE TITLE
Greet the user with their preferred name

### ISSUE THREAD

**Issue by RubenVerborgh** [SolidLabResearch/Challenges#4]
## Pitch
- Users like to be **addressed by their preferred name** within apps.
  - This is a name that they are known as, or use for themselves.
- In some cultures, this is the first name.
  - It could also be a user-chosen name or nickname.
- Other cultures do not have the concept of a first name, so another strategy is needed.

## Desired solution
- Encode the rules for arriving at the preferred name in a declarative way.
  - For instance, using reasoning rules.
- Expose the functionality via JavaScript
  - For instance, through `getUserPreferredName` or `user.preferredName`

## Acceptance criteria
- I want to see an example app with at the top right a user bar.
- This user bar shows a profile picture, and "Hi, Alex".
  - Here, "Alex" is the logged-in user's preferred name.
- This demo should work for people from different backgrounds.
  - people without a first name
  - people with a nickname
  - people with a chosen name

## Pointers
- https://www.w3.org/International/questions/qa-personal-names
- https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/

## Scenarios
- https://github.com/SolidLabResearch/Challenges/issues/5

**Comment by pietercolpaert** [comment]
I wonder whether we also have to consider language negotiation here, depending on whether you’re e.g., browsing a page in Chinese or in English?

**Comment by RubenVerborgh** [comment]
Absolutely. But perhaps we should have language as a separate challenge, and tie the two together later.

**Comment by RubenVerborgh** [comment]
Recently saw a good example

![Screenshot_20220525-203320](https://user-images.githubusercontent.com/675313/170565982-ece4d5de-edfe-45fc-84a0-0d1ec7c77869.png)

So this, without asking (every time).

**Comment by pheyvaer** [comment]
@Sindhu-Vasireddy What is the status of this challenge?

**Comment by Sindhu-Vasireddy** [comment]
https://github.com/Sindhu-Vasireddy/SolidLoginGreeter is deployed at https://sindhu-vasireddy.github.io/SolidLoginGreeter/ as a demo for this challenge

**Comment by RubenVerborgh** [comment]
Great, we're off to a good start!

Now I propose to specifically zoom in on this bit:

> - This demo should work for people from different backgrounds.
>   - people without a first name
>   - people with a nickname
>   - people with a chosen name

This it not fully reflected in https://github.com/Sindhu-Vasireddy/SolidLoginGreeter/blob/5b73ebea3775a6c84964c229d061fde3c16d3a5d/main.js#L116-L141 yet

And let's also implement this in a configurable way. So basically, that there is a declarative file in which I can state those preferences, something like this (pseudo-code, do not use):

```
preferred-name:
1. foaf:nick
2. foaf:givenName
3. schema:givenName
4. schema:fullName
```

So basically, there should be a generic piece of code that does not know about things like `foaf:givenName`; all of that knowledge should be in a declarative JSON or RDF file.

**Comment by Sindhu-Vasireddy** [comment]
Update: The user can now specify custom preference rules (and modify the order in default rules as well) to choose which name predicate (using eye reasoner on browser) to consider for the name greeting on the app, there is a [default ruleset](https://github.com/Sindhu-Vasireddy/SolidLoginGreeter/blob/main/PreferenceRules.n3) using foaf and schema vocabulary added to user pod in case the user doesn't want to specify the rules and just use the app. PFA the Reasoning proof result n3 file for a sample result. 
![ReasoningProof_n3](https://user-images.githubusercontent.com/98752594/188286530-019937d7-1e8e-4e9e-bdd9-10f5ff820292.png)
Please let me know your feedback @RubenVerborgh .

**Comment by pheyvaer** [comment]
@Sindhu-Vasireddy The link https://github.com/Sindhu-Vasireddy/SolidLoginGreeter/blob/main/PreferenceRules.n3 gives a 404 for me.

**Comment by Sindhu-Vasireddy** [comment]
@pheyvaer I have updated the link now it is in https://github.com/Sindhu-Vasireddy/SolidLoginGreeter/blob/main/public/PreferenceRules.n3. The file was earlier there but I updated the code in the latest commit and forgot to update the link here.

**Comment by RubenVerborgh** [comment]
Great work, thanks @Sindhu-Vasireddy!

**Comment by pheyvaer** [comment]
You find the report for this challenge [here](https://github.com/SolidLabResearch/Challenges/blob/main/reports/greet-user-with-their-preferred-name.md).
